### PR TITLE
refactor: helpers copied between apps get one owner each

### PR DESCRIPTION
**Pull request by jan-kubica**
## Summary

Helpers that had been copied between the api and web apps get one owner each, one commit per helper.

- `@stll/user-agent`: the two byte-identical `parse-user-agent.ts` copies.
- `@stll/collation`: the bounded-LRU web version wins, plus the api's `compareCodepoint`; 38 call sites. `require-cached-collator` now points at the package.
- `@stll/stable-stringify`: the api semantics (cycles, bigint, symbol, function) for both apps.
- `@stll/time` gains `addDays`, `parseIsoDateLocal`, `isIsoDateString`, `isoDateParts`; the api keeps its decision-date bounds and UTC cursor helpers. `no-raw-date-parsing` text updated.
- The api brands ids with the contract's `toSafeId` instead of restating it.
- The api's option-colour validator derives from one `as const` tuple that names the design-system copy; a shared package was not introduced because the published design system carries no workspace runtime dependency and the api must not import a React kit.
- `template-conditions` `isRecord` deliberately accepts arrays (a dotted path segment may be an array index); now tested and commented.

Ratchets: `cross-app-lib-path-copies` 28 to 22, `cross-workspace-duplicate-export-names` 66 to 60, api lib bucket 291 to 289, web lib bucket 143 to 139. Changeset: `@stll/template-conditions` patch.

## Verification

Every moved test file and the directly affected consumers, ratchet `--self-test`/`--check`, `bun run lint:ws`, the fixture lint, the changeset guard. Lint and typecheck run in CI.

## Summary by CodeRabbit

* **New Features**
  * Added shared utilities for locale-aware sorting, deterministic value serialisation, date handling and user-agent parsing.
  * Improved dotted-path resolution for array indexes in template conditions.
* **Bug Fixes**
  * Date handling now consistently supports calendar-day arithmetic, time zones and daylight-saving transitions.
  * Technical keys receive stable, environment-independent ordering.
* **Documentation**
  * Added package documentation and clarified date, sorting and template-condition behaviour.
* **Tests**
  * Expanded coverage for date edge cases, array path resolution, serialisation, sorting and user-agent detection.

### CODE REVIEW

File: apps/api/src/db/schema-validators.ts [hunk 1]
```diff
@@ -16,7 +16,16 @@ import {
...
+/**
+ * The sixteen preset colours a select option may carry.
+ *
+ * `@stll/ui`'s `option-color` module holds the same list, because it maps each
+ * name to a CSS custom property. The kit is published and deliberately carries
+ * no `@stll/*` runtime dependency, and this app must not import a React kit, so
+ * neither side can derive from the other today; a name added here has to be
+ * added there too, or the token resolves to the empty colour.
+ */
+export const NAMED_OPTION_COLORS = [
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/db/schema-validators.ts]: Implemented in 094b5772ad (rebased head 67b3705b51): the tuple is module-private.

File: packages/collation/src/collation.ts [hunk 2]
```diff
@@ -79,3 +82,13 @@ export const compareByLocale = (
...
+
+/**
+ * Plain codepoint comparator for `.sort()` / `.toSorted()` on strings that
+ * are ids, paths, or other technical keys rather than user-facing text (lock
+ * ordering, archive entry order, search tiebreaks, ...). Unlike
+ * `.localeCompare()`, this is bit-identical across environments regardless
+ * of runtime/ICU locale.
+ */
+export const compareCodepoint = (a: string, b: string): number =>
+  a < b ? -1 : Number(a > b);
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/collation/src/collation.ts]: Not changing the order: this PR moves the comparator verbatim from the api, and lock orderings and archive entry orders were computed under it. The order is UTF-16 code-unit order, the same as JavaScript `<` and the default sort; the doc comment and README now say so precisely, and a test pins the supplementary-character case (U+1F600 before U+E000). A rename is a separate change across 29 call sites.

File: packages/stable-stringify/src/stable-stringify.test.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import { stableStringify } from "./stable-stringify";
+
+describe("stableStringify", () => {
+  test("is insensitive to the order keys were assembled in", () => {
+    expect(stableStringify({ b: 1, a: 2 })).toBe(
+      stableStringify({ a: 2, b: 1 }),
+    );
+  });
+
+  test("orders keys by codepoint, not by locale", () => {
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/stable-stringify/src/stable-stringify.test.ts]: Not changing the order: fingerprints computed by this function are persisted (loop-detector signatures, review finding identities), and this PR moves the implementation verbatim. The header now states UTF-16 code-unit order (the order of JavaScript `<`), and a test pins the supplementary-plane key case.

File: packages/stable-stringify/src/stable-stringify.ts [hunk 4]
```diff
@@ -2,8 +2,19 @@
...
- * Shared by the chat loop detector's tool-call signatures and the tool defect
- * memo's tool+args keys, which must agree on what "the same call" means.
+ * Callers compare fingerprints — the chat loop detector's tool-call
+ * signatures, a review finding's identity, an editor's dirty check — so two
+ * structurally equal values must stringify identically no matter which order
+ * their keys were assembled in.
+ *
+ * Values JSON has no form for are given one rather than dropped: `undefined`
+ * (including an explicitly-undefined key, so it stays distinguishable from an
+ * absent one), `bigint`, `symbol`, and functions. A cycle serializes as
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/stable-stringify/src/stable-stringify.ts:45]: Documented rather than widened: the contract is JSON-shaped input (primitives, arrays, plain objects), which is what every caller fingerprints after a JSON boundary. The header and README now say that a `Date`, `Map`, or `Set` reads as `{}`, and a test pins it. Widening the contract would change persisted fingerprints, so it is not part of a move.
